File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteMiddle(head *ListNode) *ListNode {
    n := 0
    curr := head
    for curr != nil {
        n++
        curr = curr.Next
    }

    if n == 2 {
        head.Next = nil
        return head
    } else if n == 1 {
        return nil
    }

    mid := n / 2
    prev := head
    for i := 0; i < mid - 1; i++ {
        prev = prev.Next
    }

    prev.Next = prev.Next.Next

    return head
}

// linked list
// time: O(n)
// space: O(1)

File: 2095-delete-the-middle-node-of-a-linked-list/2095-delete-the-middle-node-of-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        n = 0
        curr = head
        while curr:
            n += 1
            curr = curr.next

        if n == 1:
            return None
        
        mid = n // 2
        curr = head
        for i in range(mid - 1):
            curr = curr.next
        
        curr.next = curr.next.next

        return head

# linked list
# time: O(n)
# space: O(1)